cpp/src/rolling/rolling.cu

```diff
@@ -0,0 +1,260 @@
+ * @param in_col[in]  Pointers to input column's data
+ * @param in_cols_valid[in]  Pointers to the validity mask of the input column
+ * @param window[in]  The static rolling window size, accumulates from
+ *                in_col[i-window+1] to in_col[i] inclusive
+ * @param min_periods[in]  Minimum number of observations in window required to
+ *                have a value, otherwise 0 is stored in the valid bit mask
+ * @param forward_window[in]  The static rolling window size in the forward
+ *                direction, accumulates from in_col[i] to
+ *                in_col[i+forward_window] inclusive
+ * @param[in] window_col The window size values, window_col[i] specifies window
+ *                size for element i. If window_col = NULL, then window is used as
+ *                the static window size for all elements
+ * @param[in] min_periods_col The minimum number of observation values,
+ *                min_periods_col[i] specifies minimum number of observations for
+ *                element i. If min_periods_col = NULL, then min_periods is used as
+ *                the static value for all elements
+ * @param[in] forward_window_col The forward window size values,
+ *                forward_window_col[i] specifies forward window size for element i.
+ *                If forward_window_col = NULL, then forward_window is used as the
+ *                static forward window size for all elements
+
+ */
+template <typename ColumnType, template <typename AggType> class agg_op, bool average>
+__global__
+void gpu_rolling(gdf_size_type nrows,
+		 ColumnType *out_col, gdf_valid_type *out_col_valid,
+		 ColumnType *in_col, gdf_valid_type *in_col_valid,
+		 gdf_size_type window,
+		 gdf_size_type min_periods,
+		 gdf_size_type forward_window,
+		 const gdf_size_type *window_col,
+		 const gdf_size_type *min_periods_col,
+		 const gdf_size_type *forward_window_col)
+{
+  using MaskType = uint32_t;
+  constexpr uint32_t BITS_PER_MASK{sizeof(MaskType) * 8};
+
+  gdf_size_type i = blockIdx.x * blockDim.x + threadIdx.x;
+  gdf_size_type stride = blockDim.x * gridDim.x;
+
+  agg_op<ColumnType> op;
+
+  auto active_threads = __ballot_sync(0xffffffff, i < nrows);
+  while(i < nrows)
+  {
+    ColumnType val = agg_op<ColumnType>::IDENTITY;
+    volatile gdf_size_type count = 0;	// declare this as volatile to avoid the compiler optimizing the counter
+
+    // dynamic window handling
+    if (window_col != NULL) window = window_col[i];
+    if (min_periods_col != NULL) min_periods = max(min_periods_col[i], 1);	// at least one observation is required
+    if (forward_window_col != NULL) forward_window = forward_window_col[i];
+
+    // compute bounds
+    gdf_size_type start_index = max((gdf_size_type)0, i - window + 1);
+    gdf_size_type end_index = min(nrows, i + forward_window + 1);       // exclusive
+
+    // aggregate
+    for (size_t j = start_index; j < end_index; j++) {
+      bool const input_is_valid{gdf_is_valid(in_col_valid, j)};
```

Comment:
I'm not sure why reading the mask elements as 1B vs 4B would make a difference since you get the whole cache line anyways.

But if you'd like, you can modify the kernel to accept `bit_mask_t*` for the `in/out_col_valid` (you'd need to `reinterpret_cast` before calling the kernel). `bit_mask_t` is a 4B type. Then you can use https://github.com/rapidsai/cudf/blob/branch-0.7/cpp/src/bitmask/bit_mask.cuh#L159

Reply:
It's probably just a matter of having 4x more loads...